module/state_synchronization/indexer/execution.go

```diff
@@ -0,0 +1,207 @@
+package indexer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+	"golang.org/x/exp/maps"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/onflow/flow-go/cmd/util/ledger/migrations"
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/executiondatasync/execution_data"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+// ExecutionState indexes the execution state.
+type ExecutionState struct {
+	registers  storage.RegisterIndex
+	headers    storage.Headers
+	events     storage.Events
+	indexRange *SequentialIndexRange
+	log        zerolog.Logger
+}
+
+// New execution state indexer used to ingest block execution data and index it by height.
+// The passed RegisterIndex storage must be populated to include the first and last height otherwise the indexer
+// won't be initialized to ensure we have bootstrapped the storage first.
+func New(
+	log zerolog.Logger,
+	registers storage.RegisterIndex,
+	headers storage.Headers,
+	events storage.Events,
+) (*ExecutionState, error) {
+	// we must have the first height set in the storage otherwise error out
+	first, err := registers.FirstHeight()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to initialize the indexer with missing first height")
+	}
+
+	// we must have the latest height set in the storage otherwise error out
+	last, err := registers.LatestHeight()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to initialize the indexer with missing latest height")
+	}
+
+	indexRange, err := NewSequentialIndexRange(first, last)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debug().
+		Uint64("first_height", first).
+		Uint64("latest_height", last).
+		Msgf("initialized indexer range")
+
+	return &ExecutionState{
+		registers:  registers,
+		headers:    headers,
+		events:     events,
+		indexRange: indexRange,
+		log:        log.With().Str("component", "execution_indexer").Logger(),
+	}, nil
+}
+
+// RegisterValues retrieves register values by the register IDs at the provided block height.
+// Even if the register wasn't indexed at the provided height, returns the highest height the register was indexed at.
+// Expected errors:
+// - storage.ErrNotFound if the register by the ID was never indexed
+// - ErrIndexBoundary if the height is out of indexed height boundary
+func (i *ExecutionState) RegisterValues(IDs flow.RegisterIDs, height uint64) ([]flow.RegisterValue, error) {
+	_, err := i.indexRange.Contained(height)
```

Comment:
I think this check is duplicated, since we already have such check [in the storage layer](https://github.com/onflow/flow-go/pull/4714/files#diff-1a7907c8e806ab5cf121da9cba610bcb23a62e6902e41552d6f961eae3383d0fR62)

We keep only a single source of truth, which is the storage to decide whether a certain height has been indexed.

Reply:
I don't understand this decision or the direction of the storage layer implementing the domain logic of the indexer. I think it's up to the indexer module foremost to make sure the data is correctly formatted, validated, and handled. Putting this logic into the storage layer will cause the following issues:
- possible duplication of implementation if we move away from pebble and want to implement different storage (we just did now for badger so not unlikely), 
- testing the indexer requires storage implementation to validate working, no unit tests possible, requires integration with db, makes testing harder in general,
- the storage API doesn't imply these checks are being performed,

Going in this direction indexer module is not needed at all, we could also transform the data types in the storage layer and do all of the remaining parts of the indexer. We would also have no nice way of testing this besides setting up a database to carry out these domain logic tests, this in my opinion is always a code-smell, when testing logic requires integration tests. I'm a bliever in domain-driven design and that's why I think this breaks it in a big way. This falls directly under active record design which has known downsides (some which I mentioned above): https://en.wikipedia.org/wiki/Active_record_pattern

> Another critique of the active record pattern is that, due to the strong coupling of database interaction and application logic, an active record object does not follow the [single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle) and [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns). This is opposed to [multitier architecture](https://en.wikipedia.org/wiki/Multitier_architecture), which properly addresses these practices.

cc @peterargue

Reply:
If we had the storage layer (storage.RegisterInex) as you mentioned which only provides API to store registers, update latest height, then it’s not clear who owns the responsibility of making sure registers and latest height are updated together. Since registers and latest height have to be updated in a single batch, then the storage API would have to expose the batch API for to the caller (Indexer), not to mention that the batch API has to be database agnostic.

If we move the check out of the storage layer, then we have to make sure the RegisterIndexer is the only user of the storage layer, no one should use the storage layer directly, otherwise getting a register for a height is supposed to return a register updated at a lower height, but actually would return not found.

If RegisterIndexer has to be the only user of the storage layer, then they could be merged, which is the current storage layer (storage.RegisterIndex). 

I think what’s making it special here is the way we store and query registers, we have to update in sequentially increased height order, and getting a correct value relies on the assumption of the store method being called correctly.

I also don’t like duplicating the checks (checking if height is stored in strictly incremental order), which I pointed out, because it makes debugging harder. In case the user of the Indexer get a wrong latest height, then we would have to suspect both the Indexer and storage layer, since the check are duplicated. 

Another note: Storehouse would like to reuse the same storage.RegisterIndex to store and get registers, if the check is not there, it has to implement it again, duplicating the check again. If we kept the check in the Indexer, which also indexes events, transaction results etc that are only needed by AN, RN, then the Indexer can’t be reused by EN storehouse. 

I think the simplest solution is to keep the check in storage.RegisterIndex, which can be reused by both AN and EN.

Reply:
I understand and as I said I can make it the way you said it, I won't oppose too much. 
But anyway let me give you one last description of how I see my design play out:

You have a register indexer module that contains all the domain logic ensuring sequential heights, converting of registers, and anything else that is related to indexing registers. 
You can then have an execution state indexer which includes this register indexer. All indexers could be defined by an interface that has `Index(value T, height uint64) error` method. It also contains other indexers (for events, results etc). 
The storage layer for each contains the `Store(value T, height uint64)` method, which in batch updates both value and height. Batch update is pebble specific, maybe some other dbs will have transactions, my point is all that is abstracted away. 
If you want to index registers then yes, you use it via the indexer module, I guess that is true for any other model, if you use db directly you might corrupt the state anyway. 
Why I think this design has advantages: 
- it allows each indexer to own it's own domain, now all other indexed values like events rely on heights being stored as part of the register storage, and I don't think that's clean, what if you only want to index events in some other parts of code
- storehouse would just use the register index module and won't have to reimplement anything
- it's easy to test and doesn't require database setup to test 
- if you want to query, and you are out of range, the database doesn't need to know about it, it only knows it doesn't have the required value and returns not found, the indexer itself can interpret the request as out of range

> If RegisterIndexer has to be the only user of the storage layer, then they could be merged, which is the current storage layer (storage.RegisterIndex).
I don't think that's true in general, you might often have a single use for a component but you don't merge them because that violates a single-responsibility pattern - for example, here, one responsibility is to persist data, the other is to validate it.

Again, this is not to push back, I will make the changes, is just that I feel my design has benefits and I wanted to get them across even if we don't go with it.

Reply:
I see your point.

What I'm asking for is that:
1.  we have a module that implements the `RegisterIndex` interface, which persists and validates the data. So that it can be reused by both AN and EN.
2. In the implementation, the validation is not duplicated.